modules/core/src/test/java/org/apache/ignite/internal/util/VersatileReadWriteLockTest.java

```diff
@@ -0,0 +1,606 @@
+    }
+
+    @ParameterizedTest
+    @EnumSource(BlockingWriteLockAcquisition.class)
+    void writeLockDoesNotAllowReadLockToBeAcquired(BlockingWriteLockAcquisition acquisition) {
+        acquisition.acquire(lock);
+
+        assertThatReadLockAcquireAttemptBlocksForever();
+
+        lock.writeUnlock();
+    }
+
+    private void assertThatReadLockAcquireAttemptBlocksForever() {
+        assertThatActionBlocksForever(lock::readLock);
+    }
+
+    @ParameterizedTest
+    @EnumSource(BlockingWriteLockAcquisition.class)
+    void writeLockDoesNotAllowReadLockToBeAcquiredBySameThread(BlockingWriteLockAcquisition acquisition) {
+        assertThatActionBlocksForever(() -> {
+            acquisition.acquire(lock);
+            lock.readLock();
+        });
+
+        lock.writeUnlock();
+    }
+
+    @CartesianTest
+    @EnumSource(BlockingWriteLockAcquisition.class)
+    void writeLockDoesNotAllowWriteLockToBeAcquired(
+            @Enum(BlockingWriteLockAcquisition.class) BlockingWriteLockAcquisition firstAttempt,
+            @Enum(BlockingWriteLockAcquisition.class) BlockingWriteLockAcquisition secondAttempt
+    ) {
+        firstAttempt.acquire(lock);
+
+        assertThatWriteLockAcquireAttemptBlocksForever(secondAttempt);
+
+        lock.writeUnlock();
+    }
+
+    @CartesianTest
+    @EnumSource(BlockingWriteLockAcquisition.class)
+    void writeLockDoesNotAllowWriteLockToBeAcquiredBySameThread(
+            @Enum(BlockingWriteLockAcquisition.class) BlockingWriteLockAcquisition firstAttempt,
+            @Enum(BlockingWriteLockAcquisition.class) BlockingWriteLockAcquisition secondAttempt
+    ) {
+        assertThatActionBlocksForever(() -> {
+            firstAttempt.acquire(lock);
+            secondAttempt.acquire(lock);
+        });
+
+        lock.writeUnlock();
+    }
+
+    @Test
+    void readUnlockReleasesTheLock() throws Exception {
+        lock.readLock();
+        lock.readUnlock();
+
+        runWithTimeout(lock::writeLock);
+    }
+
+    @ParameterizedTest
+    @EnumSource(BlockingWriteLockAcquisition.class)
+    void writeUnlockReleasesTheLock(BlockingWriteLockAcquisition acquisition) throws Exception {
+        acquisition.acquire(lock);
+        lock.writeUnlock();
+
+        assertThatReadLockCanBeAcquired();
+    }
+
+    @Test
+    void shouldNotAllowInterleavingHoldingReadAndWriteLocks() {
+        lock.writeLock();
+
+        assertFalse(lock.tryReadLock());
+
+        lock.writeUnlock();
+
+        lock.readLock();
+
+        assertFalse(lock.tryWriteLock());
+
+        lock.readUnlock();
+
+        // Test that we can operate with write locks now.
+        lock.writeLock();
+        lock.writeUnlock();
+    }
+
+    @Test
+    void readLockReleasedLessTimesThanAcquiredShouldStillBeTaken() throws Exception {
+        lock.readLock();
+
+        Future<?> future = executor.submit(() -> {
+            lock.readLock();
+            lock.readUnlock();
+        });
+        future.get(10, SECONDS);
```

Comment:
Usually we use something like `willCompleteSuccessfully`. You can get a completable future instead of a future if you call `CompletableFuture.runAsync(runnable, executor)`

Reply:
I see, thank you!